### PR TITLE
Use flags for auth settings

### PR DESCRIPTION
**Pull request by tempoz**
Moves the auth config into auth, the selfauth flag into selfauth, and the saml config into saml.
Biggest change is that auth config is now managed and retrieved through the Authenticator object,
which required interface changes, and that as a result, a nil authenticator is no longer a valid
state. Adds Register functions to move implementation details out of main.

---

**Version bump**: Minor

### CODE REVIEW

File: enterprise/server/auth/auth.go [hunk 1]
```diff
@@ -469,23 +484,24 @@ func newOpenIDAuthenticator(ctx context.Context, env environment.Env, oauthProvi
 		return nil, err
 	}
 
-	// Set the JWT key.
-	jwtKey = []byte(env.GetConfigurator().GetAuthJWTKey())
-
 	// Initialize API Key -> Group cache unless it's disabled by config.
 	var akgCache *apiKeyGroupCache
-	if env.GetConfigurator().GetAuthAPIKeyGroupCacheTTL() != "0" {
+	if *apiKeyGroupCacheTTL != time.Duration(0) {
```
Comment (vadimberezniker) [inline enterprise/server/auth/auth.go:489]: This is an (unintentional) change in behavior. 
The cache is on by default and disabled if the config value is set to "0" (vs the default config value of "").
Now that we have default flag values you could set the default value of apiKeyGroupCacheTTL to 5 minutes and get rid of defaultAPIKeyGroupCacheTTL

File: enterprise/server/auth/auth.go [hunk 2]
```diff
@@ -35,6 +39,20 @@ import (
...
+	enableAnonymousUsage = flag.Bool("auth.enable_anonymous_usage", false, "If true, unauthenticated build uploads will still be allowed but won't be associated with your organization.")
+	oauthProviders       = []config.OauthProvider{}
+	jwtKey               = flag.String("auth.jwt_key", "set_the_jwt_in_config", "The key to use when signing JWT tokens.")
+	apiKeyGroupCacheTTL  = flag.Duration("auth.api_key_group_cache_ttl", 0, "Override for the TTL for API Key to Group caching. Set to '0' to disable cache.")
+	httpsOnlyCookies     = flag.Bool("auth.https_only_cookies", false, "If true, cookies will only be set over https connections.")
+	disableRefreshToken  = flag.Bool("auth.disable_refresh_token", false, "If true, the offline_access scope which requests refresh tokens will not be requested.")
+)
+
+func init() {
+	flagutil.StructSliceVar(&oauthProviders, "auth.oauth_providers", "")
```
Comment (vadimberezniker) [inline enterprise/server/auth/auth.go:53]: Can you add a doc string?

File: enterprise/server/auth/auth.go [hunk 3]
```diff
@@ -704,9 +782,8 @@ func ClaimsFromSubID(env environment.Env, ctx context.Context, subID string) (*C
 	// role within the configured admin group, set their authenticated user to
 	// *only* have access to the org being impersonated.
 	if c := requestcontext.ProtoRequestContextFromContext(ctx); c != nil && c.GetImpersonatingGroupId() != "" {
-		adminGroupID := env.GetConfigurator().GetAuthAdminGroupID()
 		for _, membership := range claims.GetGroupMemberships() {
-			if membership.GroupID != adminGroupID || membership.Role != role.Admin {
+			if membership.GroupID != env.GetAuthenticator().AdminGroupID() || membership.Role != role.Admin {
```
Comment (vadimberezniker) [inline enterprise/server/auth/auth.go:786]: Can we use `a.adminGroupID` here?

File: server/nullauth/nullauth.go [hunk 4]
```diff
@@ -9,7 +9,45 @@ import (
...
+
+func (a *NullAuthenticator) AdminGroupID() string {
+	return a.adminGroupID
+}
+
+func (a *NullAuthenticator) AnonymousUsageEnabled() bool {
+	return !a.anonymousUsageDisabled
+}
+
+func (a *NullAuthenticator) EnableAnonymousUsage() {
```
Comment (vadimberezniker) [inline server/nullauth/nullauth.go:32]: Are these setters (enable, disable, set) used anywhere?

**Review by vadimberezniker — CHANGES_REQUESTED** [review]

File: enterprise/server/auth/auth.go (hunk 3 above)
Comment (tempoz) [reply enterprise/server/auth/auth.go:780]: Done!

File: server/nullauth/nullauth.go (hunk 4 above)
Comment (tempoz) [reply server/nullauth/nullauth.go]: Not anymore, apparently; removed.
Thanks for the catch!

File: enterprise/server/auth/auth.go [hunk 5]
```diff
@@ -35,6 +39,21 @@ import (
 	oidc "github.com/coreos/go-oidc"
 )
 
+var (
+	adminGroupID         = flag.String("auth.admin_group_id", "", "ID of a group whose members can perform actions only accessible to server admins.")
+	enableAnonymousUsage = flag.Bool("auth.enable_anonymous_usage", false, "If true, unauthenticated build uploads will still be allowed but won't be associated with your organization.")
+	oauthProviders       = []config.OauthProvider{}
+	jwtKey               = flag.String("auth.jwt_key", "set_the_jwt_in_config", "The key to use when signing JWT tokens.")
+	apiKeyGroupCacheTTL  = flag.Duration("auth.api_key_group_cache_ttl", time.Duration(5)*time.Minute, "Override for the TTL for API Key to Group caching. Set to '0' to disable cache.")
```
Comment (vadimberezniker) [inline enterprise/server/auth/auth.go:47]: 5 * time.Minute should be sufficient

Comment (vadimberezniker) [inline enterprise/server/auth/auth.go:47]: Let's change "Override for the TTL for..." to "TTL for...".

**Review by vadimberezniker — APPROVED** [review]

File: enterprise/server/auth/auth.go (hunk 5 above)
Comment (tempoz) [reply enterprise/server/auth/auth.go]:
```suggestion
	apiKeyGroupCacheTTL  = flag.Duration("auth.api_key_group_cache_ttl", 5*time.Minute, "Override for the TTL for API Key to Group caching. Set to '0' to disable cache.")
```

Comment (tempoz) [inline enterprise/server/auth/auth.go]:
```suggestion
	apiKeyGroupCacheTTL  = flag.Duration("auth.api_key_group_cache_ttl", time.Duration(5)*time.Minute, "TTL for API Key to Group caching. Set to '0' to disable cache.")
```

File: server/config/config.go [hunk 6]
```diff
@@ -230,8 +230,8 @@ type cacheConfig struct {
 
 type authConfig struct {
 	JWTKey               string          `yaml:"jwt_key" usage:"The key to use when signing JWT tokens."`
-	APIKeyGroupCacheTTL  string          `yaml:"api_key_group_cache_ttl" usage:"Override for the TTL for API Key to Group caching. Set to '0' to disable cache."`
-	OauthProviders       []OauthProvider `yaml:"oauth_providers"`
+	APIKeyGroupCacheTTL  time.Duration   `yaml:"api_key_group_cache_ttl" usage:"Override for the TTL for API Key to Group caching. Set to '0' to disable cache."`
```
Comment (tempoz) [inline server/config/config.go:233]:
```suggestion
	APIKeyGroupCacheTTL  time.Duration   `yaml:"api_key_group_cache_ttl" usage:"TTL for API Key to Group caching. Set to '0' to disable cache."`
```

File: enterprise/server/auth/auth.go (hunk 5 above)
Comment (tempoz) [inline enterprise/server/auth/auth.go]:
```suggestion
	apiKeyGroupCacheTTL  = flag.Duration("auth.api_key_group_cache_ttl", time.Duration(5)*time.Minute, "TTL for API Key to Group caching. Set to '0' to disable cache.")
```

**Pull request merged**